xbmc/cores/AudioEngine/Utils/PackerMAT.cpp

```diff
@@ -0,0 +1,440 @@
+  // Write MAT middle marker, if needed
+  // The MAT middle marker always needs to be in the exact same spot, any audio data will be split.
+  // If we're currently writing padding, then the marker will be considered as padding data and
+  // reduce the amount of padding still required.
+  if (GetCount() <= MAT_POS_MIDDLE && GetCount() + size > MAT_POS_MIDDLE)
+  {
+    // write as much data before the middle code as we can
+    int nBytesBefore = MAT_POS_MIDDLE - GetCount();
+    AppendData(data, nBytesBefore, type);
+    remaining -= nBytesBefore;
+
+    // write the MAT middle code
+    AppendData(mat_middle_code.data(), mat_middle_code.size(), Type::DATA);
+
+    // if we're writing padding, deduct the size of the code from it
+    if (type == Type::PADDING)
+      remaining -= mat_middle_code.size();
+
+    // write remaining data after the MAT marker
+    if (remaining > 0)
+      remaining = FillDataBuffer(data + nBytesBefore, remaining, type);
+
+    return remaining;
+  }
+
+  // not enough room in the buffer to write all the data,
+  // write as much as we can and add the MAT footer
+  if (GetCount() + size >= MAT_BUFFER_LIMIT)
+  {
+    // write as much data before the middle code as we can
+    int nBytesBefore = MAT_BUFFER_LIMIT - GetCount();
+    AppendData(data, nBytesBefore, type);
+    remaining -= nBytesBefore;
+
+    // write the MAT end code
+    AppendData(mat_end_code.data(), mat_end_code.size(), Type::DATA);
+
+    assert(GetCount() == MAT_BUFFER_SIZE);
+
+    // MAT markers don't displace padding, so reduce the amount of padding
+    if (type == Type::PADDING)
+      remaining -= mat_end_code.size();
+
+    // any remaining data will be written in future calls
+    return remaining;
+  }
+
+  AppendData(data, size, type);
+
+  return 0;
+}
+
+void CPackerMAT::FlushPacket()
+{
+  if (GetCount() == 0)
+    return;
+
+  assert(GetCount() == MAT_BUFFER_SIZE);
+
+  // normal number of samples per frame
+  const uint16_t frameSamples = 40 << (m_state.ratebits & 7);
+  const uint32_t nMATSamples = (frameSamples * 24);
+
+  // push MAT packet to output queue
+  m_outputQueue.emplace_back(std::move(m_buffer));
+
+  if (m_outputQueue.size() > 1)
+    CLog::LogF(LOGDEBUG,
+               "several MAT packets generated in a row, the size of the output queue is {}",
+               m_outputQueue.size());
+
+  // we expect 24 frames per MAT frame, so calculate an offset from that
+  // this is done after delivery, because it modifies the duration of the frame,
+  //  eg. the start of the next frame
+  if (nMATSamples != m_state.nSamples)
+    m_state.nSamplesOffset += m_state.nSamples - nMATSamples;
+
+  m_state.nSamples = 0;
+
+  m_buffer.clear();
+  m_bufferCount = 0;
+}
+
+TrueHDMajorSyncInfo CPackerMAT::ParseTrueHDMajorSyncHeaders(const uint8_t* p, int buffsize)
+{
+  TrueHDMajorSyncInfo info;
+
+  if (buffsize < 32)
+    return {};
+
+  // parse major sync and look for a restart header
+  int majorSyncSize = 28;
+  if (p[29] & 1)
```

Comment:
What's the significance of index 29? And 1?

Reply:
added code comment